File: module2/class1/exemplo14.cpp
/*
================================================================================
ATIVIDADE PRÁTICA 14 - MEMORY FRAGMENTATION (C++)
================================================================================

OBJETIVO:
- Demonstrar fragmentação de memória com alocações de tamanhos variados
- Usar Memory profiler para identificar padrões de fragmentação
- Otimizar usando memory pools ou allocators customizados
- Medir impacto da fragmentação na performance

PROBLEMA:
- Alocações/dealocações aleatórias fragmentam o heap
- Fragmentação reduz localidade de cache
- Memory profiler mostrará heap fragmentado

SOLUÇÃO:
- Usar memory pools para objetos de tamanho similar
- Alocação em blocos contíguos melhora cache locality

================================================================================
*/

#include <iostream>
#include <chrono>
#include <vector>
#include <random>
#include <algorithm>
using namespace std;

void demonstrateMemoryFragmentation() {
    cout << "Starting memory fragmentation demonstration..." << endl;
    cout << "Monitor Memory profiler - should see fragmented allocation pattern" << endl;
    
    const int ALLOCATIONS = 10000;
    vector<void*> allocatedBlocks;
    
    random_device rd;
    mt19937 gen(rd());
    uniform_int_distribution<> sizeDist(16, 4096); // Random sizes 16B to 4KB
    
    auto start = chrono::high_resolution_clock::now();
    
    // Phase 1: Allocate random sized blocks
    for (int i = 0; i < ALLOCATIONS; i++) {
        int size = sizeDist(gen);
        void* ptr = malloc(size);
        if (ptr) {
            allocatedBlocks.push_back(ptr);
            // Write to memory to ensure it's actually used
            memset(ptr, i % 256, size);
        }
        
        if (i % 1000 == 0) {
            cout << "Allocated " << i << "/" << ALLOCATIONS << " random-sized blocks..." << endl;
        }
    }
    
    // Phase 2: PERFORMANCE ISSUE - Random deallocation creates fragmentation
    shuffle(allocatedBlocks.begin(), allocatedBlocks.end(), gen);
    
    // Free every other block to create holes
    for (size_t i = 0; i < allocatedBlocks.size(); i += 2) {
        free(allocatedBlocks[i]);
        allocatedBlocks[i] = nullptr;
    }
    
    // Phase 3: Try to allocate large contiguous blocks
    vector<void*> largeBlocks;
    const int LARGE_SIZE = 8192; // 8KB blocks
    
    for (int i = 0; i < 100; i++) {
        void* ptr = malloc(LARGE_SIZE);
        if (ptr) {
            largeBlocks.push_back(ptr);
            memset(ptr, 0xFF, LARGE_SIZE);
        } else {
            cout << "Failed to allocate large block " << i << " due to fragmentation!" << endl;
        }
    }
    
    auto end = chrono::high_resolution_clock::now();
    auto duration = chrono::duration_cast<chrono::milliseconds>(end - start);
    
    cout << "Memory fragmentation test completed in: " << duration.count() << " ms" << endl;
    cout << "Random allocations made: " << ALLOCATIONS << endl;
    cout << "Large blocks successfully allocated: " << largeBlocks.size() << "/100" << endl;
    
    // Cleanup
    for (void* ptr : allocatedBlocks) {
        if (ptr) free(ptr);
    }
    for (void* ptr : largeBlocks) {
        if (ptr) free(ptr);
    }
}

int main() {
    cout << "Starting memory fragmentation demonstration..." << endl;
    cout << "Task: Creating fragmented heap with random allocations" << endl;
    cout << "Monitor Memory Usage Tool for fragmentation patterns" << endl << endl;
    
    demonstrateMemoryFragmentation();
    
    cout << endl << "=== PROFILING ANALYSIS ===" << endl;
    cout << "Check Memory profiler for:" << endl;
    cout << "- Fragmented heap layout" << endl;
    cout << "- Difficulty allocating large contiguous blocks" << endl;
    cout << "- Reduced memory efficiency" << endl;
    
    return 0;
}

/*
================================================================================
VERSÃO CORRIGIDA (DESCOMENTE PARA USAR MEMORY POOL)
================================================================================

#include <iostream>
#include <chrono>
#include <vector>
#include <array>
using namespace std;

// CORREÇÃO: Simple memory pool for fixed-size allocations
template<size_t BlockSize, size_t BlockCount>
class MemoryPool {
private:
    alignas(max_align_t) char memory[BlockSize * BlockCount];
    bool used[BlockCount] = {false};
    
public:
    void* allocate() {
        for (size_t i = 0; i < BlockCount; i++) {
            if (!used[i]) {
                used[i] = true;
                return &memory[i * BlockSize];
            }
        }
        return nullptr; // Pool exhausted
    }
    
    void deallocate(void* ptr) {
        if (!ptr) return;
        
        char* charPtr = static_cast<char*>(ptr);
        if (charPtr >= memory && charPtr < memory + sizeof(memory)) {
            size_t index = (charPtr - memory) / BlockSize;
            if (index < BlockCount) {
                used[index] = false;
            }
        }
    }
};

void demonstrateMemoryPool() {
    cout << "Starting memory pool demonstration..." << endl;
    cout << "Monitor Memory profiler - should see contiguous allocation pattern" << endl;
    
    const int ALLOCATIONS = 5000;
    
    // CORREÇÃO: Memory pools for different sizes reduce fragmentation
    MemoryPool<64, 2000> smallPool;   // 64-byte blocks
    MemoryPool<256, 2000> mediumPool; // 256-byte blocks
    MemoryPool<1024, 1000> largePool; // 1KB blocks
    
    vector<void*> smallBlocks, mediumBlocks, largeBlocks;
    
    auto start = chrono::high_resolution_clock::now();
    
    // Phase 1: Allocate from pools - no fragmentation
    for (int i = 0; i < ALLOCATIONS / 3; i++) {
        void* small = smallPool.allocate();
        void* medium = mediumPool.allocate();
        void* large = largePool.allocate();
        
        if (small) {
            memset(small, i % 256, 64);
            smallBlocks.push_back(small);
        }
        if (medium) {
            memset(medium, i % 256, 256);
            mediumBlocks.push_back(medium);
        }
        if (large) {
            memset(large, i % 256, 1024);
            largeBlocks.push_back(large);
        }
        
        if (i % 500 == 0) {
            cout << "Pool allocated " << i << " blocks of each size..." << endl;
        }
    }
    
    // Phase 2: Deallocate some blocks - no fragmentation due to pool structure
    for (size_t i = 0; i < smallBlocks.size(); i += 2) {
        smallPool.deallocate(smallBlocks[i]);
    }
    
    // Phase 3: Reallocate - efficient reuse of pool memory
    int reallocated = 0;
    for (int i = 0; i < 500; i++) {
        void* ptr = smallPool.allocate();
        if (ptr) {
            reallocated++;
            memset(ptr, 0xAA, 64);
        }
    }
    
    auto end = chrono::high_resolution_clock::now();
    auto duration = chrono::duration_cast<chrono::milliseconds>(end - start);
    
    cout << "Memory pool test completed in: " << duration.count() << " ms" << endl;
    cout << "Pool allocations made: " << ALLOCATIONS << endl;
    cout << "Successful reallocations: " << reallocated << "/500" << endl;
    cout << "No fragmentation due to pool-based allocation" << endl;
}

int main() {
    cout << "Starting optimized memory pool demonstration..." << endl;
    cout << "Task: Using memory pools to prevent fragmentation" << endl;
    cout << "Monitor Memory Usage Tool for contiguous allocation patterns" << endl << endl;
    
    demonstrateMemoryPool();
    
    cout << endl << "=== OPTIMIZATION RESULTS ===" << endl;
    cout << "Improvements:" << endl;
    cout << "- No heap fragmentation due to pool allocation" << endl;
    cout << "- Better memory locality and cache performance" << endl;
    cout << "- Predictable allocation/deallocation performance" << endl;
    cout << "- Reduced memory overhead" << endl;
    
    return 0;
}

================================================================================
*/

File: module2/class1/exemplo14.cs
/*
================================================================================
ATIVIDADE PRÁTICA 14 - LARGE OBJECT HEAP (LOH) PRESSURE (C#)
================================================================================

OBJETIVO:
- Demonstrar pressure no Large Object Heap (LOH) com objetos >85KB
- Usar Memory profiler para identificar LOH allocations e GC impact
- Otimizar usando object pooling e ArrayPool<T>
- Medir impacto do LOH na performance do GC

PROBLEMA:
- Objetos >85KB vão para LOH que não é compactado frequentemente
- Muitas alocações LOH causam GC pressure e fragmentação
- Memory profiler mostrará LOH growth e Gen 2 collections

SOLUÇÃO:
- Usar ArrayPool<T> para reutilizar large arrays
- Object pooling para large objects

================================================================================
*/

using System;
using System.Buffers;
using System.Diagnostics;

class Program {
    static void DemonstrateLOHPressure() {
        Console.WriteLine("Starting Large Object Heap pressure demonstration...");
        Console.WriteLine("Monitor Memory profiler - should see LOH allocations and Gen 2 GC");
        
        const int ITERATIONS = 1000;
        const int LARGE_ARRAY_SIZE = 90000; // >85KB goes to LOH
        
        var sw = Stopwatch.StartNew();
        long initialMemory = GC.GetTotalMemory(false);
        int initialGen2Collections = GC.CollectionCount(2);
        
        for (int i = 0; i < ITERATIONS; i++) {
            // PERFORMANCE ISSUE: Creating large arrays that go to LOH
            byte[] largeArray = new byte[LARGE_ARRAY_SIZE]; // Goes to Large Object Heap
            
            // Use the array to prevent it from being optimized away
            for (int j = 0; j < 1000; j++) {
                largeArray[j] = (byte)(i % 256);
            }
            
            // Array goes out of scope - LOH pressure builds up
            
            if (i % 100 == 0) {
                long currentMemory = GC.GetTotalMemory(false);
                int currentGen2Collections = GC.CollectionCount(2);
                Console.WriteLine($"LOH allocation {i}/{ITERATIONS}:");
                Console.WriteLine($"  Memory: {currentMemory:N0} bytes");
                Console.WriteLine($"  Gen 2 collections: {currentGen2Collections}");
            }
        }
        
        sw.Stop();
        long finalMemory = GC.GetTotalMemory(false);
        int finalGen2Collections = GC.CollectionCount(2);
        
        Console.WriteLine($"LOH pressure test completed in: {sw.ElapsedMilliseconds} ms");
        Console.WriteLine($"Memory increase: {(finalMemory - initialMemory):N0} bytes");
        Console.WriteLine($"Gen 2 collections triggered: {finalGen2Collections - initialGen2Collections}");
        Console.WriteLine($"Large arrays allocated: {ITERATIONS} (each {LARGE_ARRAY_SIZE:N0} bytes)");
    }
    
    static void Main() {
        Console.WriteLine("Starting Large Object Heap demonstration...");
        Console.WriteLine("Task: Creating many large arrays (>85KB)");
        Console.WriteLine("Monitor Memory Usage Tool for LOH pressure and GC impact");
        Console.WriteLine();
        
        DemonstrateLOHPressure();
        
        Console.WriteLine();
        Console.WriteLine("=== PROFILING ANALYSIS ===");
        Console.WriteLine("Check Memory profiler for:");
        Console.WriteLine("- Large Object Heap allocations");
        Console.WriteLine("- Increased Gen 2 garbage collections");
        Console.WriteLine("- Memory pressure and GC pauses");
    }
}

/*
================================================================================
VERSÃO CORRIGIDA (DESCOMENTE PARA USAR ARRAYPOOL)
================================================================================

using System;
using System.Buffers;
using System.Diagnostics;

class Program {
    static void DemonstrateArrayPool() {
        Console.WriteLine("Starting ArrayPool demonstration...");
        Console.WriteLine("Monitor Memory profiler - should see reduced LOH pressure");
        
        const int ITERATIONS = 1000;
        const int LARGE_ARRAY_SIZE = 90000;
        
        var sw = Stopwatch.StartNew();
        long initialMemory = GC.GetTotalMemory(false);
        int initialGen2Collections = GC.CollectionCount(2);
        
        // CORREÇÃO: Use ArrayPool to reuse large arrays
        var pool = ArrayPool<byte>.Shared;
        
        for (int i = 0; i < ITERATIONS; i++) {
            // CORREÇÃO: Rent from pool instead of allocating new array
            byte[] largeArray = pool.Rent(LARGE_ARRAY_SIZE); // Reuses existing arrays
            
            try {
                // Use the array
                for (int j = 0; j < 1000; j++) {
                    largeArray[j] = (byte)(i % 256);
                }
                
                // Process data...
            }
            finally {
                // CORREÇÃO: Return to pool for reuse
                pool.Return(largeArray); // Array goes back to pool, not GC
            }
            
            if (i % 100 == 0) {
                long currentMemory = GC.GetTotalMemory(false);
                int currentGen2Collections = GC.CollectionCount(2);
                Console.WriteLine($"ArrayPool usage {i}/{ITERATIONS}:");
                Console.WriteLine($"  Memory: {currentMemory:N0} bytes");
                Console.WriteLine($"  Gen 2 collections: {currentGen2Collections}");
            }
        }
        
        sw.Stop();
        long finalMemory = GC.GetTotalMemory(false);
        int finalGen2Collections = GC.CollectionCount(2);
        
        Console.WriteLine($"ArrayPool test completed in: {sw.ElapsedMilliseconds} ms");
        Console.WriteLine($"Memory increase: {(finalMemory - initialMemory):N0} bytes");
        Console.WriteLine($"Gen 2 collections triggered: {finalGen2Collections - initialGen2Collections}");
        Console.WriteLine($"Arrays reused from pool: {ITERATIONS}");
    }
    
    static void Main() {
        Console.WriteLine("Starting optimized ArrayPool demonstration...");
        Console.WriteLine("Task: Reusing large arrays with ArrayPool<T>");
        Console.WriteLine("Monitor Memory Usage Tool for reduced LOH pressure");
        Console.WriteLine();
        
        DemonstrateArrayPool();
        
        Console.WriteLine();
        Console.WriteLine("=== OPTIMIZATION RESULTS ===");
        Console.WriteLine("Improvements:");
        Console.WriteLine("- Dramatically reduced LOH allocations");
        Console.WriteLine("- Fewer Gen 2 garbage collections");
        Console.WriteLine("- Lower memory pressure");
        Console.WriteLine("- Array reuse eliminates allocation overhead");
        Console.WriteLine("- Better GC performance overall");
    }
}

================================================================================
*/
